### PR TITLE
feat!: Returns index of failing opcode and transformation mapping

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Related to the display of source location for failing opcodes.
ACVM solve now returns the index of the failing opcode, and tracks modifications done to the opcodes during the compilation process in a separate vector

Resolves 

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: acvm/src/pwg/mod.rs [hunk 1]
```diff
@@ -102,6 +102,9 @@ pub struct ACVM<B: BlackBoxFunctionSolver> {
     /// Note that this doesn't include any opcodes which are waiting on a pending foreign call.
     opcodes: Vec<Opcode>,
 
+    /// the position of the each opcode in the opcodes
```
Comment (kevaundray) [inline acvm/src/pwg/mod.rs:105]:
```suggestion
    /// The position of the each opcode in the opcodes
```

File: acir/src/circuit/mod.rs [hunk 2]
```diff
@@ -54,6 +54,10 @@ impl Circuit {
         let circuit = rmp_serde::from_slice(buf_d.as_slice()).unwrap();
         Ok(circuit)
     }
+
+    pub fn default_opcode_idx(&self) -> Vec<usize> {
```
[the remark this answers is not in the document]
Comment (guipublic) [reply acir/src/circuit/mod.rs]: Done

File: acvm/src/pwg/mod.rs [hunk 3]
```diff
@@ -102,21 +102,31 @@ pub struct ACVM<B: BlackBoxFunctionSolver> {
     /// Note that this doesn't include any opcodes which are waiting on a pending foreign call.
     opcodes: Vec<Opcode>,
 
+    /// The position of the each opcode in the opcodes
+    opcodes_idx: Vec<OpcodeLabel>,
```
Comment (jfecher) [inline acvm/src/pwg/mod.rs:106]: Can we remove this field and change opcodes to:
```
opcodes: Vec<(Opcode, OpcodeLabel)>,
```

File: acir/src/circuit/brillig.rs [hunk 4]
```diff
@@ -30,3 +30,23 @@ pub struct Brillig {
...
+        serialize_vector.extend(rmp_serde::to_vec(&self.outputs).unwrap());
+        serialize_vector.extend(rmp_serde::to_vec(&self.bytecode).unwrap());
+        serialize_vector.extend(rmp_serde::to_vec(&self.predicate).unwrap());
+
+        use std::collections::hash_map::DefaultHasher;
+        use std::hash::Hasher;
+
+        let mut hasher = DefaultHasher::new();
+        hasher.write(&serialize_vector);
+        hasher.finish()
```
[the remark this answers is not in the document]
Comment (kevaundray) [reply acir/src/circuit/brillig.rs]: This is mainly due to the bytecode not being hashable

File: acvm/src/pwg/mod.rs (hunk 3 above)
Comment (guipublic) [reply acvm/src/pwg/mod.rs]: Done

**Comment by kevaundray** [comment]
Changing this to breaking, since it modifies the error variant in OpcoderesolutionError

**Review by kevaundray — APPROVED** [review]

**Pull request merged**